server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -867,47 +874,75 @@ public void finalizeSnapshot(final SnapshotId snapshotId,
         // when writing the index-${N} to each shard directory.
         final Consumer<Exception> onUpdateFailure =
             e -> listener.onFailure(new SnapshotException(metadata.name(), snapshotId, "failed to update snapshot in repository", e));
-        final ActionListener<SnapshotInfo> allMetaListener = new GroupedActionListener<>(
-            ActionListener.wrap(snapshotInfos -> {
-                assert snapshotInfos.size() == 1 : "Should have only received a single SnapshotInfo but received " + snapshotInfos;
-                final SnapshotInfo snapshotInfo = snapshotInfos.iterator().next();
-                getRepositoryData(ActionListener.wrap(existingRepositoryData -> {
-                    final RepositoryData updatedRepositoryData =
-                        existingRepositoryData.addSnapshot(snapshotId, snapshotInfo.state(), shardGenerations);
+
+        final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
+
+        getRepositoryData(ActionListener.wrap(existingRepositoryData -> {
```

Comment:
Note to reviewers: There's no big change here. Just moved out the loading of repository data to be the first step so that we have it available for the index metadata writer threads for the lookup and removed its lazy loading from `allMetaListener` below, all other changes are just indent changes.